File: csharp/Library/Book.cs
namespace SammanCoaching.Library;
public sealed class Book
{
    private Book(
        string title,
        Isbn isbn,
        string copyId,
        Member? borrowedBy,
        Member? reservedBy,
        DateOnly? borrowedOn,
        DateOnly? dueDate)
    {
        Title = RequireText(title, nameof(title));
        Isbn = isbn ?? throw new ArgumentNullException(nameof(isbn));
        CopyId = RequireText(copyId, nameof(copyId));
        BorrowedBy = borrowedBy;
        ReservedBy = reservedBy;
        BorrowedOn = borrowedOn;
        DueDate = dueDate;
    }
    public string Title { get; }
    public Isbn Isbn { get; }
    public string CopyId { get; }
    public Member? BorrowedBy { get; }
    public Member? ReservedBy { get; }
    public DateOnly? DueDate { get; }

    private DateOnly? BorrowedOn { get; }
    public static Book Available(string title, Isbn isbn, string copyId) =>
        AvailableCopy(title, isbn, copyId, null);
    public Book ReservedByMember(Member member) =>
        new(Title, Isbn, CopyId, BorrowedBy, member ?? throw new ArgumentNullException(nameof(member)), BorrowedOn, DueDate);
    public Book Borrow(Member member) => Borrow(member, LoanPeriod.DefaultPeriod());
    public Book Borrow(Member member, LoanPeriod loanPeriod)
    {
        ArgumentNullException.ThrowIfNull(member);
        ArgumentNullException.ThrowIfNull(loanPeriod);

        if (IsBorrowed())
        {
            throw new InvalidOperationException("Book is already borrowed");
        }

        if (ReservedBy is not null && ReservedBy != member)
        {
            throw new InvalidOperationException("Book is reserved by another reader");
        }

        var today = DateOnly.FromDateTime(DateTime.Today);
        return new Book(Title, Isbn, CopyId, member, null, today, today.AddDays(loanPeriod.Days));
    }
    public Book ReturnBook(Member member)
    {
        if (!IsBorrowed())
        {
            throw new InvalidOperationException("Book is not borrowed");
        }

        if (BorrowedBy != (member ?? throw new ArgumentNullException(nameof(member))))
        {
            throw new InvalidOperationException("Book is borrowed by another reader");
        }

        return AvailableCopy(Title, Isbn, CopyId, ReservedBy);
    }
    public Book Renew(Member member, LoanPeriod loanPeriod)
    {
        ArgumentNullException.ThrowIfNull(loanPeriod);

        if (!IsBorrowed())
        {
            throw new InvalidOperationException("Book is not borrowed");
        }

        if (BorrowedBy != (member ?? throw new ArgumentNullException(nameof(member))))
        {
            throw new InvalidOperationException("Book is borrowed by another reader");
        }

        if (ReservedBy is not null && ReservedBy != member)
        {
            throw new InvalidOperationException("Book is reserved by another reader");
        }

        var renewedDueDate = DueDate!.Value.AddDays(loanPeriod.Days);
        if (renewedDueDate > BorrowedOn!.Value.AddDays(LoanPeriod.MaxDays))
        {
            throw new InvalidOperationException($"Book cannot be borrowed longer than {LoanPeriod.MaxDays} days");
        }

        return new Book(Title, Isbn, CopyId, BorrowedBy, ReservedBy, BorrowedOn, renewedDueDate);
    }
    public bool IsBorrowed() => BorrowedBy is not null;

    private static Book AvailableCopy(string title, Isbn isbn, string copyId, Member? reservedBy) =>
        new(title, isbn, copyId, null, reservedBy, null, null);

    private static string RequireText(string value, string fieldName)
    {
        ArgumentNullException.ThrowIfNull(value, fieldName);
        if (string.IsNullOrWhiteSpace(value))
        {
            throw new ArgumentException($"{fieldName} must not be blank", fieldName);
        }

        return value;
    }
}

File: java/lib/src/main/java/org/sammancoaching/library/Book.java
package org.sammancoaching.library;

import java.time.LocalDate;
import java.util.Objects;
import java.util.Optional;
public final class Book {
    private final String title;
    private final Isbn isbn;
    private final String copyId;
    private final Member borrowedBy;
    private final Member reservedBy;
    private final LocalDate borrowedOn;
    private final LocalDate dueDate;

    private Book(
            String title,
            Isbn isbn,
            String copyId,
            Member borrowedBy,
            Member reservedBy,
            LocalDate borrowedOn,
            LocalDate dueDate
    ) {
        this.title = requireText(title, "title");
        this.isbn = Objects.requireNonNull(isbn, "isbn");
        this.copyId = requireText(copyId, "copyId");
        this.borrowedBy = borrowedBy;
        this.reservedBy = reservedBy;
        this.borrowedOn = borrowedOn;
        this.dueDate = dueDate;
    }
    public static Book available(String title, Isbn isbn, String copyId) {
        return availableCopy(title, isbn, copyId, null);
    }

    private static Book availableCopy(String title, Isbn isbn, String copyId, Member reservedBy) {
        return new Book(title, isbn, copyId, null, reservedBy, null, null);
    }
    public Book reservedBy(Member member) {
        return new Book(title, isbn, copyId, borrowedBy, Objects.requireNonNull(member, "member"), borrowedOn, dueDate);
    }
    public Book borrow(Member member) {
        return borrow(member, LoanPeriod.defaultPeriod());
    }
    public Book borrow(Member member, LoanPeriod loanPeriod) {
        Objects.requireNonNull(member, "member");
        Objects.requireNonNull(loanPeriod, "loanPeriod");
        if (isBorrowed()) {
            throw new IllegalStateException("Book is already borrowed");
        }

        if (reservedBy != null && !reservedBy.equals(member)) {
            throw new IllegalStateException("Book is reserved by another reader");
        }

        LocalDate today = LocalDate.now();
        return new Book(title, isbn, copyId, member, null, today, today.plusDays(loanPeriod.days()));
    }
    public Book returnBook(Member member) {
        if (!isBorrowed()) {
            throw new IllegalStateException("Book is not borrowed");
        }

        if (!borrowedBy.equals(Objects.requireNonNull(member, "member"))) {
            throw new IllegalStateException("Book is borrowed by another reader");
        }

        return availableCopy(title, isbn, copyId, reservedBy);
    }
    public Book renew(Member member, LoanPeriod loanPeriod) {
        Objects.requireNonNull(loanPeriod, "loanPeriod");
        if (!isBorrowed()) {
            throw new IllegalStateException("Book is not borrowed");
        }

        if (!borrowedBy.equals(Objects.requireNonNull(member, "member"))) {
            throw new IllegalStateException("Book is borrowed by another reader");
        }

        if (reservedBy != null && !reservedBy.equals(member)) {
            throw new IllegalStateException("Book is reserved by another reader");
        }

        LocalDate renewedDueDate = dueDate.plusDays(loanPeriod.days());
        if (renewedDueDate.isAfter(borrowedOn.plusDays(LoanPeriod.MAX_DAYS))) {
            throw new IllegalStateException("Book cannot be borrowed longer than " + LoanPeriod.MAX_DAYS + " days");
        }

        return new Book(title, isbn, copyId, borrowedBy, reservedBy, borrowedOn, renewedDueDate);
    }
    public String title() {
        return title;
    }
    public Isbn isbn() {
        return isbn;
    }
    public String copyId() {
        return copyId;
    }
    public boolean isBorrowed() {
        return borrowedBy != null;
    }
    public Optional<Member> borrowedBy() {
        return Optional.ofNullable(borrowedBy);
    }
    public Optional<Member> reservedBy() {
        return Optional.ofNullable(reservedBy);
    }
    public LocalDate dueDate() {
        return dueDate;
    }

    private static String requireText(String value, String fieldName) {
        Objects.requireNonNull(value, fieldName);
        if (value.isBlank()) {
            throw new IllegalArgumentException(fieldName + " must not be blank");
        }
        return value;
    }
}
